lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs

```diff
@@ -2036,6 +1973,42 @@ balanceTransactionSpec = describe "balanceTransaction" $ do
                     ]
             tx `shouldBe` Left ErrBalanceTxMaxSizeLimitExceeded
 
+    describe "effect of maxValSize" $ do
+        it "change split when needed" $ property $ \(room :: Small Int) -> do
+            -- Test what happens when the maxValSize pparam is set close to the
+            -- size of the only utxo in a wallet. For extra difficulty, we let
+            -- the token bundle have a coin-value which will increase in size by
+            -- 4 bytes when increased.
+            let walletBundle =
+                    (generateWith (GenSeed 0) (GenSize 100) genTokenBundle)
+                    { TokenBundle.coin = Coin $ 2 ^ (32 :: Int) - 1 }
+            let pparams =
+                    let bundleSize = unTxSize $
+                            computeTokenBundleSerializedLengthBytes walletBundle
+                        maxValSize' = fromMaybe (error "maxValSize': underflow")
+                                $ intCastMaybe @Integer @Natural
+                                $ intCast bundleSize + intCast room
+                    in
+                        Write.withConstraints RecentEraBabbage $ pparamsLedger
+                            (mockPParamsForBalancing @Cardano.BabbageEra)
+                                & ppMaxValSizeL .~ maxValSize'
+
+            let res = balanceTx @Cardano.BabbageEra
+                    (mkTestWallet $ utxoWithBundles [walletBundle])
+                    (Write.ProtocolParameters pparams)
+                    dummyTimeTranslation
+                    testStdGenSeed
+                    (inputPartialTx (TokenBundle.fromCoin (Coin 1_000_000)))
+
+            let costOfIncreasingCoin' = 4
+            let expectedOuts =
+                    if room - costOfIncreasingCoin' >= 0
+                    then 1
+                    else 2 -- change doesn't fit in a single output
+            case res of
+                Right tx -> length (outputs tx) === expectedOuts
+                Left e -> counterexample (show e) $ property False
```

Comment:
@Anviking what's the motivation for adding this test?

To check that change bundles are split correctly, we already have a comprehensive set of tests within `Cardano.CoinSelection`. Is there any benefit to repeating this test coverage here?

Reply:
I added it for additional e2e coverage when I was considering not keeping the tests now in `TokenBundleSizeSpec`.

I will remove it now, but I have a feeling that tests like this could be very important in the future if we are able to use the ledger rules as ground truth.